### ISSUE TITLE
Check for `// SPDX-License-Identifier: XXX` in all source files

### ISSUE THREAD

**Issue by tgross35** [rust-lang/rust-clippy#11663]
### What it does

Some projects require a license be specified in all source files. Clippy could check for this, just make sure the comment is present as the first line.

By default it could just check for presence of the line in all files and verify it is a valid license with the help of the crate `spdx`. A config option could specify allowed or required licenses.

### Example

```rust
// SPDX-License-Identifier: MIT OR Apache-2.0
```

**Comment by 90degs2infty** [comment]
You're probably aware of this, but just in case you're not: there is e.g. the [reuse software](https://reuse.software/) project maintaining a [specification](https://reuse.software/spec/) of attaching licensing information to source code files in a language/toolstack-independent way. It might be beyond scope, but maybe consider not only checking for the `SPDX-License-Identifier` tags but instead check for compliance with above specification? Though I must admit that this would probably lead to a re-implementation of the [`reuse` tool](https://github.com/fsfe/reuse-tool) (at least the `reuse lint` command).

One more thought on the `SPDX-License-Identifier` tag: I haven't delved into the spec too deep, but to my knowledge (at least, this is how I read both the reuse and the SPDX specs) the tag may refer to custom licenses like so:

```rust
// SPDX-License-Identifier: LicenseRef-CUSTOM-LICENSE-NAME
```

(see e.g. [this part of the reuse spec](https://reuse.software/spec/#license-files) for details and further references). So if `clippy` wants to verify the specified license's validity, it might have to locate the corresponding license file and verify it's presence (instead of doing a mere lookup in the [SPDX-maintained table of licenses](https://spdx.org/licenses/)).

At any rate, I'd love to see a lint going in this direction! 👍

**Comment by J-ZhengLi** [comment]
> By default it could just check for presence of the line in all files and verify it is a valid license with the help of the crate spdx. A config option could specify allowed or required licenses.

I don't know if this is possible to add in clippy, that crate would requires periodically updates in order to keep the license list up-to-date, correct?

> So if clippy wants to verify the specified license's validity, it might have to locate the corresponding license file and verify it's presence

And this might be hard for clippy, afaik, clippy don't have that much of freedom to lookup license files in the source code directory, unless... you find those file and push them into rustc_sessions when invoking the compiler (same as how clippy.toml being handled)? I'm probably wrong tho.

However, it could read the license info from `Cargo.toml` by default, or having a dedicate `expected-license-header` configuration, then lint if some **source code file** that doesn't have them at the very top.

**Comment by bbastin** [comment]
> I don't know if this is possible to add in clippy, that crate would requires periodically updates in order to keep the license list up-to-date, correct?

I don't think this would be such a huge problem, the license list does not get updated very frequently. In 2023, it saw three updates, in the three years before there were four (https://github.com/spdx/license-list-XML/releases). This could probably even be automated.

The second point seems like a larger problem, but I am not familiar with cargo internals to assess that.

**Comment by heaths** [comment]
We have a requirement to include not the SPDX license, but a full 2 lines of a copyright notice. So, more generically, we'd love to have a lint that just checks for some text (.clippy.toml or `[lints.clippy]` in Cargo.toml) in all files and recommends to add it, which some editors can take advantage of to update all supported files.

E.g. Cargo.toml
```toml
[package]
license = "MIT"

[lints.clippy]
# this would be used instead since it's more explicit
source_header = """
Copyright ...
Licensed under the MIT License.
"""
```

**Comment by heaths** [comment]
Looks like comments are counted as an `Attribute` token so we may be able to do this in an `EarlyLintPass`.

**Comment by blyxyas** [comment]
If this is going to be added, I think that @heaths approach is the best one. Any other would be full of false negatives or would be hard to discover (it would just be discovered via the lint list, as I've not seen many files that start with `SPDX-License-Identifier`).

Kinda related: Licenses take a lot of forms, for example, sqlite3 is licensed [by blessing](https://github.com/sqlite/sqlite/blob/8a894cd8043c3ddd48b79517035a4fc21e339448/src/main.c#L1).
